Fixes for the Zephyr port

Frauschi opened the pull request:
Hi there, 

This PR contains fixes for the Zephyr OS port. I implemented them while getting WolfSSL running on an STM32H573I discovery board. 
* The first commit fixes a deprecation warning in the current version of Zephyr (V 3.5).
* The second commit fixes a (probably) wrong include in wc_port.h that resulted in compile errors due to redefinition of some types and methods (this also resulted in the Zephyr WolfSSL samples all not properly compiling with the latest version).
* The third commit adds support for reading the wall clock via the Zephyr RTC interface. In case no RTC time is available (either because RTC support is disabled or no valid time is set), we fall back to the old implementation, which gives us the relative time since boot. With the RTC time, we are now able to properly validate certificates.
* The forth commit deals with the assembly optimizations for the Cortex-M architecture. When compiling with `SP_MATH` and `SP_ARM_CORTEX_M_ASM`, we get compilation errors, as the Zephyr SDK gcc compiler doesn't recognize the `asm` keyword (it needs `__asm__`). Hence, when the Zephyr port is enabled, we add a makro to map the keyword name.
* The final commit is a fix that I found in preparation for enabling the STM32 hardware acceleration inside WolfSSL along Zephyr. Even with STM32 hardware acceleration enabled, we still want to use the random number generator via the Zephyr interface (so that the Zephyr OS itself also has access to random data). For that, we have to define `NO_STM32_RNG`. This, however, resulted in not properly configuring the Zephyr port for the RNG, as the order of the configuration checks was incorrect.

All fixes have been tested using Zephyr on both a STM32H573I discovery board and a STM32H743 Nucleo board. 

Once this PR is (hopefully) merged, I will open a follow-up issue (and potentially a PR) for the STM32 hardware acceleration enablement on the STM32H5, as this doesn't work currently (both AES and HASH fails to work for me).

I'm looking forward to your feedback regarding the fixes.

dgarske commented:
Hi @Frauschi ,

Thank you for this pull request and improvements to the Zephyr port. In order to accept these we need to have a signed contributor agreement on file. Please email support@wolfssl.com with your details including information about your project and we'll provide a document to sign.

Thanks,
David Garske, wolfSSL

dgarske commented on wolfcrypt/src/wc_port.c:
```diff
@@ -3162,6 +3162,32 @@ time_t xilinx_time(time_t * timer)
+
+	/* Fallback to uptime since boot. This works for relative times, but
+	 * not for ASN.1 date validation */
     struct timespec ts;
```
Please avoid mixed declaration, so move `struct timespec ts;` to top of function. Please use `if () {` on same line.

Frauschi replied:
updated the commit, thanks

dgarske commented on wolfcrypt/src/random.c:
```diff
@@ -3623,6 +3609,20 @@ int wc_GenerateSeed(OS_Seed* os, byte* output, word32 sz)
         return ret;
     }
 
+#elif defined(WOLFSSL_SAFERTOS) || defined(WOLFSSL_LEANPSK) || \
```
I agree with this move further down. Thanks

julek-wolfssl commented:
Hi @Frauschi, great work! The zephyr github action fails to build with the following log. It uses Zephyr 3.4.0 and SDK 0.16.1. We would like to keep support for this version too.

```
[172/246] Building C object modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c.obj
FAILED: modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c.obj 
ccache /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -DKERNEL -DWOLFSSL_SETTINGS_FILE=\"user_settings-tls-generic.h\" -DWOLFSSL_USER_SETTINGS -DWOLFSSL_ZEPHYR -D__ZEPHYR__=1 -I/home/runner/work/wolfssl/wolfssl/zephyr/zephyr/include -I/home/runner/work/wolfssl/wolfssl/zephyr/twister-out/qemu_x86/zephyr/samples/wolfssl_test/sample.crypto.wolfssl_test/zephyr/include/generated -I/home/runner/work/wolfssl/wolfssl/zephyr/zephyr/soc/x86/ia32 -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfssl -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/zephyr -isystem /home/runner/work/wolfssl/wolfssl/zephyr/zephyr/lib/libc/minimal/include -isystem /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/include -isystem /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/include-fixed -m32  -fno-strict-aliasing -Werror -Os -imacros /home/runner/work/wolfssl/wolfssl/zephyr/twister-out/qemu_x86/zephyr/samples/wolfssl_test/sample.crypto.wolfssl_test/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -m32 -msoft-float -Wa,--divide --sysroot=/home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/x86_64-zephyr-elf -imacros /home/runner/work/wolfssl/wolfssl/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/zephyr/samples/wolfssl_test=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-mmx -mno-sse -D_POSIX_THREADS -std=c99 -nostdinc -MD -MT modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c.obj -MF modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c.obj.d -o modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c.obj -c /home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c
/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/random.c:3511:18: fatal error: zephyr/random/random.h: No such file or directory
 3511 |         #include <zephyr/random/random.h>
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Juliusz

julek-wolfssl commented:
@Frauschi I pushed a commit adding zephyr 3.5 to the github actions being tested. Please include this commit in the final PR.

Frauschi commented:
> Hi @Frauschi, great work! The zephyr github action fails to build with the following log. It uses Zephyr 3.4.0 and SDK 0.16.1. We would like to keep support for this version too.
> 
> Juliusz

I added an #if makro to read the autogenerated version.h header. In case we have version 3.4 or lower, we now include rand32.h, otherwise the new random.h. I think that should be the best case.

dgarske reviewed (approved):
Awaiting signed contributor agreement (tracking in ZD 16930)

dgarske commented:
Okay to test

julek-wolfssl commented:
@Frauschi Still seeing an error in the github action:

```
[200/246] Building C object modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj
FAILED: modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj 
ccache /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -DKERNEL -DWOLFSSL_SETTINGS_FILE=\"user_settings-tls-generic.h\" -DWOLFSSL_USER_SETTINGS -DWOLFSSL_ZEPHYR -D__ZEPHYR__=1 -I/home/runner/work/wolfssl/wolfssl/zephyr/zephyr/include -I/home/runner/work/wolfssl/wolfssl/zephyr/twister-out/qemu_x86/zephyr/samples/wolfssl_test/sample.crypto.wolfssl_test/zephyr/include/generated -I/home/runner/work/wolfssl/wolfssl/zephyr/zephyr/soc/x86/ia32 -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfssl -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/zephyr -isystem /home/runner/work/wolfssl/wolfssl/zephyr/zephyr/lib/libc/minimal/include -isystem /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/include -isystem /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/include-fixed -m32  -fno-strict-aliasing -Werror -Os -imacros /home/runner/work/wolfssl/wolfssl/zephyr/twister-out/qemu_x86/zephyr/samples/wolfssl_test/sample.crypto.wolfssl_test/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -m32 -msoft-float -Wa,--divide --sysroot=/home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/x86_64-zephyr-elf -imacros /home/runner/work/wolfssl/wolfssl/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/zephyr/samples/wolfssl_test=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-mmx -mno-sse -D_POSIX_THREADS -std=c99 -nostdinc -MD -MT modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj -MF modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj.d -o modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj -c /home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c
/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c: In function 'z_time':
/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c:3178:52: error: implicit declaration of function 'mktime' [-Werror=implicit-function-declaration]
 3178 |                                 time_t epochTime = mktime(tm_time);
      |                                                    ^~~~~~
/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c:34:1: note: 'mktime' is defined in header '<time.h>'; did you forget to '#include <time.h>'?
   33 |     #include <wolfssl/wolfcrypt/ecc.h>
  +++ |+#include <time.h>
   34 | #endif
cc1: all warnings being treated as errors
```

Frauschi commented:
> Awaiting signed contributor agreement (tracking in ZD 16930)

Just sent it via E-Mail.

Frauschi commented:
> @Frauschi Still seeing an error in the github action:
> 
> ```
> [200/246] Building C object modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj
> FAILED: modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj 
> ccache /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -DKERNEL -DWOLFSSL_SETTINGS_FILE=\"user_settings-tls-generic.h\" -DWOLFSSL_USER_SETTINGS -DWOLFSSL_ZEPHYR -D__ZEPHYR__=1 -I/home/runner/work/wolfssl/wolfssl/zephyr/zephyr/include -I/home/runner/work/wolfssl/wolfssl/zephyr/twister-out/qemu_x86/zephyr/samples/wolfssl_test/sample.crypto.wolfssl_test/zephyr/include/generated -I/home/runner/work/wolfssl/wolfssl/zephyr/zephyr/soc/x86/ia32 -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfssl -I/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/zephyr -isystem /home/runner/work/wolfssl/wolfssl/zephyr/zephyr/lib/libc/minimal/include -isystem /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/include -isystem /home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/include-fixed -m32  -fno-strict-aliasing -Werror -Os -imacros /home/runner/work/wolfssl/wolfssl/zephyr/twister-out/qemu_x86/zephyr/samples/wolfssl_test/sample.crypto.wolfssl_test/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -m32 -msoft-float -Wa,--divide --sysroot=/home/runner/work/wolfssl/wolfssl/zephyr-sdk-0.16.1/x86_64-zephyr-elf/x86_64-zephyr-elf -imacros /home/runner/work/wolfssl/wolfssl/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/zephyr/samples/wolfssl_test=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/runner/work/wolfssl/wolfssl/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-mmx -mno-sse -D_POSIX_THREADS -std=c99 -nostdinc -MD -MT modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj -MF modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj.d -o modules/wolfssl/CMakeFiles/..__modules__crypto__wolfssl__zephyr.dir/home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c.obj -c /home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c
> /home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c: In function 'z_time':
> /home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c:3178:52: error: implicit declaration of function 'mktime' [-Werror=implicit-function-declaration]
>  3178 |                                 time_t epochTime = mktime(tm_time);
>       |                                                    ^~~~~~
> /home/runner/work/wolfssl/wolfssl/zephyr/modules/crypto/wolfssl/wolfcrypt/src/wc_port.c:34:1: note: 'mktime' is defined in header '<time.h>'; did you forget to '#include <time.h>'?
>    33 |     #include <wolfssl/wolfcrypt/ecc.h>
>   +++ |+#include <time.h>
>    34 | #endif
> cc1: all warnings being treated as errors
> ```

Strange, I will look at it tomorrow.

embhorn commented:
@Frauschi is approved as a wolfSSL contributor.

Frauschi commented:
I updated the RTC commit once again, now (hopefully) properly handling all configuration cases. I missed that RTC (and time handling support) needs at Picolibc or newlib and doesn't work with the minimal libc. Therefore, I added checks for that, too. In case no RTC support is enabled (with is true for the qemu test), the new code is completely disabled now.

dgarske commented:
Retest this please

dgarske commented on wolfcrypt/src/sp_cortexm.c:
```diff
@@ -60,6 +60,9 @@
 #define __asm__        __asm
 #define __volatile__   volatile
 #endif
+#ifdef WOLFSSL_ZEPHYR
```
Can you tell me the error this resolves? Is it the new `register sp_digit* r asm ("r0") = (sp_digit*)r_p;` type expressions? I think we may want to fix this differently (not just for Zephyr).
Note: This file is generated, so we must also update an internal script.

Frauschi replied:
Yes, these declarations all fail using the Zehpyr SDK. I'm not sure currently whether other compilers are also affected (e.g., the `arm-none-eabi` toolchain). 
Maybe it's better to generally change the declarations to use `__asm__` instead of `asm`? I didn't realize that the file is generated. Hence, I'm not sure if I can do such a fix?

dgarske replied:
Yeah we'll need to fix this issue on our side. Feel free to keep it for now. We've got a PR open to address similar issues with IAR, so we will integrate a proper update there... #6943.

dgarske replied:
Hi @Frauschi , this has been fixed in #6943. Please revert this change and confirm 6943 solves your compilation issues. Thanks

dgarske commented on wolfcrypt/src/random.c:
```diff
@@ -3506,23 +3506,16 @@ int wc_GenerateSeed(OS_Seed* os, byte* output, word32 sz)
+        #include <zephyr/random/random.h>
+    #else
         #include <zephyr/random/rand32.h>
+    #endif 
```
```8 bit, trailing spaces or tabs:
./wolfcrypt/src/random.c:3517:    #endif
weird control chars, hard tabs, CRs, trailing whitespace:
```

Frauschi replied:
fixed

dgarske commented on wolfcrypt/src/wc_port.c:
```diff
@@ -3164,6 +3164,30 @@ time_t z_time(time_t * timer)
+    }
+    #endif
+
+	/* Fallback to uptime since boot. This works for relative times, but
```
```8 bit, trailing spaces or tabs:
./wolfcrypt/src/wc_port.c:3189:»/* Fallback to uptime since boot. This works for relative times, but
./wolfcrypt/src/wc_port.c:3190:» * not for ASN.1 date validation */
weird control chars, hard tabs, CRs, trailing whitespace:
```

Frauschi replied:
fixed

dgarske reviewed (changes requested):
Please use 4 spaces (not tabs)

Frauschi commented:
I just tried #6943 and it indeed solves the compilation issues I see with Zephyr. Hence, I deleted the related commit with my fix.

dgarske reviewed (approved):
Over to @wolfSSL-Bot

julek-wolfssl commented:
Rebased to collect fixes for some of the github actions test. For some reason `make check` now fails when compiling with `--enable-tls13 --enable-postauth`.

Frauschi commented:
> Rebased to collect fixes for some of the github actions test. For some reason `make check` now fails when compiling with `--enable-tls13 --enable-postauth`.

I can't reproduce the issue on my side. Do you think the failing check is related to my changes?

julek-wolfssl commented:
Tests passed this time. @JacobBarthelmeh should be good to go.
